### ISSUE TITLE
Prevent running projects with gdscript errors 

### ISSUE THREAD

**Issue by nullptr64** [godotengine/godot-proposals#8835]
### Describe the project you are working on

I'm learning Godot, following some tutorial project.

### Describe the problem or limitation you are having in your project

It's possible to run a project with syntactically incorrect gdscript code, the editor can highlight the incorrect line, but it doesn't stop me from running the project.
The problem is that if I don't notice the highlight, I'll have to spend time debugging why my game is not working, while it could be easily indicated by the editor. I could also not notice that something is not working and release a buggy game. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I'd like to have a toggle to enable some kind of strict mode.
In strict mode the editor would refuse to run a project with gdscript errors and would jump to the place of the first error instead. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The editor already highlights the errors (taking into account project settings where additional errors can be configured).
With the strict mode enabled, before actually running the project, the editor could iterate over all opened gdscript tabs and check if they have any errors. If yes, jump to the first found error.

Alternatively the editor could iterate over all .gd files in the project, run the existing checker on them and jump to the first found error (if any). This is more expensive, but might be better for larger project with multiple people working in parallel.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I found a reddit thread where a person asks how to do pretty much exactly what I'm proposing and another person wrote a script to load all scripts and check them somehow:
https://www.reddit.com/r/godot/comments/13gibx2/scan_whole_godot_4_project_for_gdscript_errors/

There they also say that unit testing would help. I haven't tried, so I don't know how well it would work, the proposed strict mode would also be helpful for prototyping (i.e. before it even makes sense to write tests).

Alternatively I also noticed that after I run my game, if I press ctrs+s in the editor, in some cases the game crashes/breaks on the incorrect line and the editor jumps there. But this behaviour is not consistent, I couldn't figure out why it does that for some errors and seems to completely ignore the entire .gd file in other cases.

### Is there a reason why this should be core and not an add-on in the asset library?

Since static type checking, settings for what is an error vs warning vs ignore already exist, I believe it makes a lot of sense to also treat errors as actual errors and force developers to fix them before running the project instead of ignoring them (or at least give an option for such behaviour). I think that for the vast majority of developers strict mode would reduce time spent on debugging and decrease the amount of bugs in released games.

P.S. I'm an experienced C++ developer and I'd be interested in helping with implementing this, if the proposal is accepted and the implementation is not insanely difficult (I never contributed to godot yet).

**Comment by Calinou** [comment]
Only some errors are known before you run a project (and only in scripts currently open in the script editor).

That said, for errors that are known in advance by the script editor, this could help speed up project iteration. Pressing the <kbd>F5</kbd> key should directly bring you to the first error encountered instead, without running the project.

**Comment by yosoyfreeman** [comment]
Could the default behaviour be configurable? Sometimes bad syntax can cause the debugger to point you to an error in a line which is cause by another thing.  Seeing exactly the moment it crashes is a useful way to debug.

**Comment by adamscott** [comment]
What about files that aren't used in the project per se? Like, imagine that you have a `error.gd` file with syntax errors that is not used by any node nor script. Would it prevent the project to be run?

**Comment by nullptr64** [comment]
> There may need to be a way to bypass this though (e.g. by pressing F5 a second time in quick succession, or holding down a modifier).

> Could the default behaviour be configurable?

I agree with that. I personally would prefer something like F5 for running with checking and Shift+F5 for running without checking. I'm not sure if F5 can be changed to something else at the moment. If yes, then that could be the way to change the default behaviour. If not, I could add a toggle in the settings near the settings for what is treated as error/warning/etc. 

>  imagine that you have a error.gd file with syntax errors that is not used by any node nor script. Would it prevent the project to be run?

This is an interesting edge case. We could add a special line in such a file to exclude it form producing any errors ever. The user could also rename it into error.txt, so the editor doesn't pick up it automatically. That would make sense if the file is used just for storing code. Overall depends on what the user was trying to achieve by having such a file. If it's just temporary not used and will be included later, then I think catching errors there is still useful. In the worst case the user could just comment out the entire file.

> And what about C# too?

I don't really know how C# is working right now in Godot. It's a strongly typed language, so I'd personally expect that my C# should be able to at least compile before the game can run (same with C++, if it doesn't compile, I don't want to run it, nothing good will happen anyway).

**Comment by nullptr64** [comment]
It turns out that the main issue was that Godot 4.1.1 wasn't correctly breaking on syntax errors. And that's why I was getting the described behavior when I could run the game, but some parts of it were simply not working. 

In Godot 4.2.1 for all cases that I tested so far it immediatelly breaks the game and jumps to the invalid line, which solves my main problem. 
From what I understood it wouldn't do that for scripts that are not loaded. I'm not sure what triggers the loading of a script and how likely it is that some errors would slip through because of that. So maybe it'd still be a good idea to scan the entire project for syntax errors, but is not as critical for me as the original issue.

**Comment by nullptr64** [comment]
> An example for of those cases, is when you are creating an add-on or some type of @ tool script that you use to create the game, but when running it. You may desire to run the game to check something, or just change what you are working on for some time.

If the code is used at least sometimes, I'd personally not want to have errors in it. 
Is there a use case when a tool has errors (that the editor has identified) and we want to keep them there instead of fixing them?

**Comment by AThousandShips** [comment]
Most errors are not discoverable on compile or at launch, especially most errors happening to experienced users, so the scope of this is limited, and is largely identifiable by paying attention in the editor, I'm not sure adding another layer of the same errors and information would be all that helpful, and would come with a lot of possible points of failure, what if there's an error in identifying an error? What if it's context dependent and the compiler doesn't identify it correctly and locks out the game?

**Comment by nullptr64** [comment]
> Most errors are not discoverable on compile or at launch, especially most errors happening to experienced users, so the scope of this is limited, and is largely identifiable by paying attention in the editor, I'm not sure adding another layer of the same errors and information would be all that helpful, and would come with a lot of possible points of failure, what if there's an error in identifying an error? What if it's context dependent and the compiler doesn't identify it correctly and locks out the game? If it's going to be low risk in that regard it would only catch a tiny portion of all possible errors, and the rest would still fire just as the game launches once it has that extra information that couldn't be known before it launches (due to things like loading data, instancing scenes, etc. that's a complex web of states and information)

I see your point. I'm coming from a background of being a C++ developer and I really enjoy having compilation errors instead of trying to debug runtime bugs. Of course also in C++ there are bugs that the compiler doesn't catch, but the more it catches the less time I spend on debugging. 

Just to reiterate, Godot 4.1.1 wasn't actually catching the errors even when I was running the project. It was simply running the game and mostly silently disabling bad parts. That was my main concern. In 4.2.1 it does show me syntax errors immediately after I run a project, which is pretty much what I wanted. 

The only remaining concern is that it only shows errors for loaded scripts. I'm not sure I personally want any errors in my code, but I can trust your judgement here if you say it's not a problem in practice.

**Comment by AThousandShips** [comment]
Well 4.1.1 is outdated so it's not a relevant point of comparison, so this proposal is mainly based on an error, thatsbee. Solved

The problem here is that you can't know at launch what scripts matter, and I don't see the value in a feature that expects *all* your code to be free of errors, I don't want to be bound to have all my code production ready at every moment, not that I need code that's intentionally broken, but I don't want to have the requirement that I either can't use this feature, or have every script in my entire project be free of errors, see the problem? Because it's impossible to know what code might be relevant, or at least very complicated and, again, error prone

This is something that could easily softlock a project with a minor bug in the editor code, some incorrect guess, that is if it doesn't cause errors when the code actually runs

Unless, again, you just want to limit this to syntax errors, which feels like it isn't all that useful as, again, most of the time you won't have syntax errors *and* they are easy to notice in the editor *and* doesn't depend on other scripts, as in the kind of errors that are harder to discover, what's the use of just a *second* error for syntax errors beyond the ones that show up on the screen?

**Comment by nullptr64** [comment]
> Unless, again, you just want to limit this to syntax errors

I'd be ok to limit this to syntax errors. Right now they feel more like warnings to me because I'm not sure there's even a difference between a warning and an error, other than the highlight color. 
Maybe there are other cases when the editor can be sure, like with type checking.

> Well 4.1.1 is outdated so it's not a relevant point of comparison, so this proposal is mainly based on an error, thatsbee. Solved

Yes, I'm ok if people close this issue. 

> I don't want to have the requirement that I either can't use this feature, or have every script in my entire project be free of errors, see the problem?

I understand that you don't want that. That's ok, the proposed strict mode doesn't have to be mandatory.
I do want such behavior. I don't see a point in running my code if there are known bugs and errors in it. I'm also not trying to convince you to change your workflow, just stating what was working well for me for many years and I think some other people could also benefit from this.

**Comment by LordMcMutton** [comment]
I'd like something like this, personally

Coming from Unity, where you were prevented from entering Play Mode (Running the project) if there were an error that would crash the game immediately.

Right now, Godot has an annoying habit of running the project, and having the game window freeze while the engine gives the error for why the game crashed. The big issue here is that it's annoying and cumbersome to have to press Close on the game engine several times to close the window, then exit out of Window's error reporting:
![image](https://github.com/godotengine/godot-proposals/assets/93609836/27b138d1-e245-44ff-bca6-e8fae9d2867d)
